modules/personalize_taxonomy_context/personalize_taxonomy_context.module

```diff
@@ -0,0 +1,78 @@
+<?php
+
+/**
+ * @file personalize_taxonomy_context.module
+ * Provides a visitor context plugin for taxonomy.
+ */
+
+/**
+ * Implements hook_personalize_visitor_contexts().
+ */
+function personalize_taxonomy_context_personalize_visitor_context() {
+  $info = array();
+  $path = drupal_get_path('module', 'personalize_taxonomy_context') . '/plugins';
+  $info['taxonomy_context'] = array(
+    'path' => $path . '/visitor_context',
+    'handler' => array(
+      'file' => 'TaxonomyContext.inc',
+      'class' => 'TaxonomyContext',
+    ),
+  );
+  return $info;
+}
+
+/**
+ * Implements hook_node_view().
+ */
+function personalize_taxonomy_context_node_view($node, $view_mode, $langcode) {
+  personalize_taxonomy_context_node_taxonomy($node, $view_mode == 'full');
+}
+
+/**
+ * Handles the node-specific taxonomy data.
+ *
+ * @param $node
+ *   (optional) If passed then the taxonomy terms for the selected node
+ *   are added to the page taxonomy context.
+ * @param bool $primary
+ *   True if this is the primary node displayed for the page, false otherwise.
+ * @return array
+ *   Returns the taxonomy context for the page.
+ */
+function personalize_taxonomy_context_node_taxonomy($node = NULL, $primary = FALSE) {
+  $taxonomy = &drupal_static(__FUNCTION__);
+  if (!isset($taxonomy)) {
+    $taxonomy = array();
+  }
+
+  // Add vocabulary and terms only if:
+  // 1) Has a node, and
+  // 2) The node is the primary node displayed for the page.
+  if (!isset($node) || !$primary) {
+    return $taxonomy;
+  }
+
+  // For each vocabulary, add any terms that are applied to the viewed node.
+  $vocabularies = array();
+  foreach($node as $property_name => $property_value) {
+    // If the field is not a vocabulary, move on.
+    if (!isset($property_value['und'])) {
```

Comment:
OK, here be dragons :/ The string 'und' represents the "undefined" language - there is actually a constant for it (LANGUAGE_NONE) which should be used instead of the 'und' string. However, we can't assume that the field language is LANGUAGE_NONE - it could actually be a language code like 'en', in which case the field array structure would be `$field_name['en'][0]...`. Have a look at how this is handled in personalize_fields module - you can probably copy some code from there: https://github.com/acquia/personalize_module/blob/7.x-1.x/modules/personalize_fields/personalize_fields.module#L244